### PR TITLE
Add surface roles

### PR DESCRIPTION
**Pull request by TheNeikos**

### CODE REVIEW

**Comment by neithernut** [comment]
ACK.

File: src/compositor/wayland/surface.c [hunk 1]
```diff
@@ -453,3 +458,22 @@ resource_destroy(
...
+    }
+
+    if (self->role && self->role != role) {
+        return -1;
+    }
+
+    self->role = role;
+
+    return 0;
+
```
Comment (matthiasbeyer) [inline src/compositor/wayland/surface.c]: Please remove this blank line.

File: src/compositor/wayland/surface.c [hunk 2]
```diff
@@ -453,3 +458,22 @@ resource_destroy(
...
+int
+ws_surface_set_role(
+    struct ws_surface* self,
+    struct wl_interface const* role
+) {
+    if (!self || !role) {
+        return -1;
+    }
+
+    if (self->role && self->role != role) {
```
Comment (matthiasbeyer) [inline src/compositor/wayland/surface.c]: A `ws_surface` is a `ws_object`, isn't it? So it should be read-locked here... I guess!

File: src/compositor/wayland/surface.c [hunk 3]
```diff
@@ -453,3 +458,22 @@ resource_destroy(
...
+        return -1;
+    }
+
+    self->role = role;
```
Comment (matthiasbeyer) [inline src/compositor/wayland/surface.c:492]: ... and write-locked here.

Comment (neithernut) [inline src/compositor/wayland/surface.c:492]: **IF** anything, the whole thing has to be write-blocked. You get deadlocks otherwise.

Comment (matthiasbeyer) [inline src/compositor/wayland/surface.c:492]: So I guess it would be feasible to state this in the documentation of this function.

**Comment by TheNeikos** [comment]
Please do tell if I can do a read lock and then a write lock

**Comment by matthiasbeyer** [comment]
On the same object? I don't think so, but I guess you should wait on @neithernut s answer here.

**Comment by matthiasbeyer** [comment]
Why would that be neccessary? A write-lock includes a read-lock!?

**Comment by TheNeikos** [comment]
Check the code, but I suppose I can do a write lock from the beginning.

**Comment by TheNeikos** [comment]
Okay, I changed the code a bit, please check that I do the locking/unlocking correctly.

File: src/compositor/wayland/surface.c [hunk 4]
```diff
@@ -453,3 +458,27 @@ resource_destroy(
...
+) {
+    int ret = -1;
+    ws_object_lock_write(&self->wl_obj.obj);
+
+    if (!self || !role) {
+        goto away;
+    }
+
+
+    if (self->role && self->role != role) {
```
Comment (matthiasbeyer) [inline src/compositor/wayland/surface.c]: What about merging this into the upper one?

At least the doubled blank line above should be removed.

File: src/compositor/wayland/surface.c [hunk 5]
```diff
@@ -453,3 +458,27 @@ resource_destroy(
...
+        goto away;
+    }
+
+    self->role = role;
+    ret = 0;
+away:
```
Comment (matthiasbeyer) [inline src/compositor/wayland/surface.c]: You really wanna call this "away"? :smile:

**Comment by matthiasbeyer** [comment]
The locking looks sane for me.

---

Edit: And btw I don't see why you would need a read lock here anyways...

**Comment by TheNeikos** [comment]
Okay, changed a few things

**Comment by matthiasbeyer** [comment]
ACK.

**Comment by neithernut** [comment]
Needs rebase.

**Comment by TheNeikos** [comment]
Rebased

**Comment by neithernut** [comment]
ACK.

**Pull request merged**